Katello/katello
# Fixes #34461 - Add table sorting

#### What are the changes introduced in this pull request?

![sortable](https://user-images.githubusercontent.com/22042343/155342871-86e1e9ab-1121-4587-a9fe-7e6f43512479.png)

1. Adds a new hook, `useTableSort`, which provides sort params for both our API and for Patternfly components.
2. Adds table sorting as per mockups to the following new Host Detail pages:
* Module streams
* Errata
* Packages (name column only)
* Repository sets
* Traces
3. Adds a new <SortableColumns /> component that replaces the `<Th>` components in the top table row.
3. Refactors the existing ContentViewsTable to use `useTableSort`. This required refactoring the entire table to use the composable TableWrapper.
4. Adds a way to surface search errors to the user. Previously, errors such as `ScopedSearch::QueryNotSupported: the field 'status' in the order statement is not valid field for search` were not shown to the user because the HTTP status code was 200. But now they can be displayed as part of a TableWrapper empty search screen.

![error_search_body](https://user-images.githubusercontent.com/22042343/155342698-7e9a5d07-6db8-4355-b0e9-11c654f18fa9.png)


#### Considerations taken when implementing this change?

There were three levels of difficulty when adding sortable table columns:
1. The field you want to sort by is already an attribute on the Rails model and is available for searching via scoped_search
2. The field you want to sort by is already an attribute on the Rails model but is not available via scoped_search
3. The field you want to sort by is rendered by a presenter (or by the front end), meaning custom sorting logic is required

Packages contains some difficulty-3 fields that I am leaving for a followup PR. Other than that, this PR covers all sortable fields required by the mockups.

For the difficulty-1 fields, the procedure was very simple:
1. Define a `const columnHeaders = ` that contains all table columns. Patternfly uses this because it really needs the table column indexes, for some reason.
2. Also define a `COLUMNS_TO_SORT_PARAMS`. This is an object containing your sortable columns. If this object includes a column, that column will show as sortable in the table. The keys are the column names and the values are the `order_by` params that are sent to the API.
3. Use the hook:
```js
const {
    pfSortParams, apiSortParams,
    activeSortColumn, activeSortDirection,
  } = useTableSort({
    allColumns: columnHeaders,
    columnsToSortParams: COLUMNS_TO_SORT_PARAMS,
    initialSortColumnName: 'Application',
  });
```
4. Modify `fetchItems` to send the `apiParams` provided by `useTableSort`
4. Add `activeSortDirection` and `activeSortColumn` to `additionalListeners`
5. Replace the `<Th>` elements in the top table row with
```js
<SortableColumnHeaders
                columnHeaders={columnHeaders}
                pfSortParams={pfSortParams}
                columnsToSortParams={COLUMNS_TO_SORT_PARAMS}
              />
```


#### What are the testing steps for this pull request?

Get a host with 
* installed packages
* errata - sync https://jlsherrill.fedorapeople.org/fake-repos/needed-errata/ or enable RHEL
* module streams (RHEL 8 or 9 required) - sync https://fixtures.pulpproject.org/rpm-with-modules/
* Traces (use the following in Rails console to create some fake traces. katello-host-tools-tracer must still be installed) Or run 'yum update' on the host and you'll probably get some real traces.
```rb
c7 = Host::Managed.find_by(name: 'centos7.fedora.example.com')
100.times { |count| ::Katello::HostTracer.create!(application: "foo-#{count}", helper: "systemctl restart foo-#{count}",
app_type: ["session", "static", "daemon"][count % 3], host_id: c7.id)}
```
* Repository sets - sync any repo

* Click around and make sure all the sorting works
* check Content Views and make sure everything still works
* do some bogus search like 'boo = hi' and view the new error messages
